zenml/core/pipelines/base_pipeline_test.py

```diff
@@ -0,0 +1,217 @@
+from zenml.core.datasources.base_datasource import BaseDatasource
+from zenml.core.datasources.image_datasource import ImageDatasource
+from zenml.core.backends.orchestrator.base.orchestrator_base_backend import \
+    OrchestratorBaseBackend
+from zenml.core.steps.base_step import BaseStep
+from zenml.core.repo.repo import Repository
+from zenml.utils.enums import PipelineStatusTypes, GDPComponent
+from zenml.core.standards import standard_keys as keys
+from zenml.utils import exceptions, path_utils
+
+ZENML_ROOT = zenml.__path__[0]
+TEST_ROOT = os.path.join(ZENML_ROOT, "testing")
+
+pipelines_dir = os.path.join(TEST_ROOT, "test_pipelines")
+repo: Repository = Repository.get_instance()
+repo.zenml_config.set_pipelines_dir(pipelines_dir)
+
+
+def test_executed():
+    name = "my_pipeline"
+    p = BasePipeline(name=name)
+    assert not p.is_executed_in_metadata_store
+
+    random_run_pipeline = random.choice(repo.get_pipelines())
+
+    assert random_run_pipeline.is_executed_in_metadata_store
+
+
+def test_naming():
+    name = "my_pipeline"
+    p = BasePipeline(name=name)
+    # file_name = p.file_name
+    pipeline_name = p.pipeline_name
+
+    assert p.get_name_from_pipeline_name(pipeline_name) == name
+
+
+def test_get_status(run_test_pipelines):
+    run_test_pipelines()
+    name = "my_pipeline"
+    p = BasePipeline(name=name)
+
+    # passing this test means that if the pipeline is not in the metadata
+    # store, it is (semantically) not started yet
+    assert p.get_status() == PipelineStatusTypes.NotStarted.name
+
+    run_pipeline = random.choice(repo.get_pipelines())
+
+    assert run_pipeline.get_status() == PipelineStatusTypes.Succeeded.name
+
+
+def test_register_pipeline(delete_config):
+    name = "my_pipeline"
+    p: BasePipeline = BasePipeline(name=name)
+
+    p._check_registered()
+
+    random_run_pipeline = random.choice(repo.get_pipelines())
+    with pytest.raises(exceptions.AlreadyExistsException):
+        random_run_pipeline._check_registered()
+
+    p.register_pipeline({"name": name})
+
+    delete_config(p.file_name)
+
+
+def test_add_datasource():
+    name = "my_pipeline"
+    p: BasePipeline = BasePipeline(name=name)
+
+    p.add_datasource(BaseDatasource(name="my_datasource"))
+
+    assert isinstance(p.datasource, BaseDatasource)
+
+    assert not p.steps_dict[keys.TrainingSteps.DATA]
+
+
+def test_pipeline_copy():
+    random_run_pipeline = random.choice(repo.get_pipelines())
+
+    new_name = "my_second_pipeline"
+
+    new_pipeline = random_run_pipeline.copy(new_name=new_name)
+
+    assert not new_pipeline._immutable
+    assert isinstance(new_pipeline, random_run_pipeline.__class__)
+
+
+def test_get_pipeline_config():
+    # test just the pipeline config block added to the steps_config
+    name = "my_pipeline"
+    p: BasePipeline = BasePipeline(name=name)
+
+    config = p.get_pipeline_config()
+
+    p_name = p.pipeline_name
+
+    assert config[keys.PipelineKeys.NAME] == p_name
```

Comment:
These keys might change a bit with the `newpipelines` branch but we can refactor in there